test/integration/replicaset/replicaset_test.go

```diff
@@ -268,250 +510,438 @@ func TestAdoption(t *testing.T) {
+
+	// Update .Spec.Replicas and verify .Status.Replicas is changed accordingly
+	scaleRS(t, c, rs, 3)
+	scaleRS(t, c, rs, 0)
+	scaleRS(t, c, rs, 2)
+
+	// Add a template annotation change to test RS's status does update
+	// without .Spec.Replicas change
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	updateRS(t, rsClient, rs, func(rs *v1beta1.ReplicaSet) {
+		rs.Spec.Template.Annotations = map[string]string{"test": "annotation"}
+	})
+	newRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("failed to update rs %s: %v", rs.Name, err)
+	}
+	if newRS.Generation <= rs.Generation {
+		t.Fatalf("Failed to verify .Generation has incremented for rs %s: %v", rs.Name, err)
+	}
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newRS, err = rsClient.Get(rs.Name, metav1.GetOptions{})
 		if err != nil {
 			return false, err
 		}
-		if updatedRS.Status.Replicas != *rs.Spec.Replicas {
-			return false, nil
-		} else {
-			return true, nil
-		}
+		return newRS.Status.ObservedGeneration > rs.Generation, nil
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to verify .Status.ObservedGeneration has incremented for rs %s: %v", rs.Name, err)
 	}
 }
 
-func TestUpdateSelectorToAdopt(t *testing.T) {
-	// We have pod1, pod2 and rs. rs.spec.replicas=1. At first rs.Selector
-	// matches pod1 only; change the selector to match pod2 as well. Verify
-	// there is only one pod left.
-	s, closeFn, rm, informers, clientSet := rmSetup(t)
+func TestDeletingAndFailedPods(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	ns := framework.CreateTestingNamespace("rs-update-selector-to-adopt", s, t)
+	ns := framework.CreateTestingNamespace("test-deleting-and-failed-pods", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
-	rs := newRS("rs", ns.Name, 1)
-	// let rs's selector only match pod1
-	rs.Spec.Selector.MatchLabels["uniqueKey"] = "1"
-	rs.Spec.Template.Labels["uniqueKey"] = "1"
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	defer close(runControllerAndInformers(rm, informers))
 
-	stopCh := make(chan struct{})
-	informers.Start(stopCh)
-	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
+	rs := newRS("rs", ns.Name, 2)
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{})
+	rs = rss[0]
+	waitRSStable(t, c, rs)
+
+	// Verify RS creates 2 pods
+	podClient := c.Core().Pods(ns.Name)
+	pods := getPods(t, podClient, labelMap())
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %d, want 2", len(pods.Items))
+	}
+
+	// Set first pod as deleting pod
+	// Set finalizers for the pod to simulate pending deletion status
+	deletingPod := &pods.Items[0]
+	updatePod(t, podClient, deletingPod, func(pod *v1.Pod) {
+		pod.Finalizers = []string{"foregroundDeletion"}
```

Comment:
The `foregroundDeletion` finalizer might actually get processed and the Pod deleted if the GC becomes enabled in the future. It would be safer to make up a finalizer that doesn't exist, so we know it will never be satisfied. I think you have to give it a subdomain like `fake.example.com`.

Reply:
Noted. How about `fake.example.com.blockDeletion`?

Reply:
There is validation checking whether a finalizer name is standard or fully qualified. It seems that having subdomain does not work.
`fake.example.com.foregroundDeletion`

Reply:
It should work with something like `fake.example.com/finalizer`.

Reply:
Oops!
I verify `fake.example.com/blockDeletion` works.